### PR TITLE
Fixes access to callback_metrics in ddp_spawn

### LINKED ISSUE #7671: Callback metrics not being populated during multi-gpu training
## 🐛 Bug

When `plt.Trainer(gpus>1, ...)`, the `callback_metrics` dictionary appears not to be populated. I've tried to both integrate Optuna and Ray Tune, and have failed with both as a result.

To ensure this was the issue, I printed the `trainer.callback_metrics` attribute:
```
(pid=144372) callback metrics are:
(pid=144372)  {}
```

It does, however, work with 1 GPU. Unfortunately for me, my datasets are graphs, and they are so large that I can only fit one into memory at a time, so the number of GPUs = batch size, and tuning with a batch size of 1 might not be very indicative. Any help much appreciated!

### Hardware
- 2x RTX 8000 GPUs

### Software
- Python 3.8.10
- PyTorch 1.7.1
- PyTorch Lightning 1.3.1
- Ray[Tune] 1.1.0
- cudatoolkit 10.2
- OS: Linux version 3.10.0-693.11.6.el7.x86_64, Red Hat 4.8.5-16
- Installed via conda
- Any other relevant information: using SLURM

## Code

I attach the LightningModule below, which uses TorchMetrics and the self.log features, as per the docs. I did (in desperation) try setting the callback_metrics dictionary myself in `validation_epoch_end()`, but that didn't work. Neither did setting `sync_dist=True` in the `self.log()` calls.

```python
import torch
import torchmetrics
import torch_geometric as pyg
import pytorch_lightning as pl

class Interaction(pyg.nn.MessagePassing):
    def __init__(self, in_edge, in_node, out_edge, out_node):
        super(Interaction, self).__init__(
            aggr='add',
            flow="source_to_target")
        self.in_edge = 2 * in_node + in_edge
        self.in_node = in_node + out_edge
        self.mlp_edge = torch.nn.Sequential(
            torch.nn.Linear(self.in_edge, out_edge, bias=True),
            torch.nn.ReLU(),
            torch.nn.Linear(out_edge, out_edge, bias=True)
        )
        self.mlp_node = torch.nn.Sequential(
            torch.nn.Linear(self.in_node, out_node, bias=True),
            torch.nn.ReLU(),
            torch.nn.Linear(out_node, out_node, bias=True)
        )

    def forward(self, x, edge_index, edge_attrs):
        return self.propagate(
            x=x,
            edge_index=edge_index,
            edge_attrs=edge_attrs
        )

    def message(self, x_i, x_j, edge_index, edge_attrs):
        recv_send = [x_i, x_j]
        if edge_attrs is not None:
            recv_send.append(edge_attrs)
        recv_send = torch.cat(recv_send, dim=1)
        self.edge_embed = self.mlp_edge(recv_send)
        return self.edge_embed

    def update(self, aggr_out, x):
        node_embed = self.mlp_node(torch.cat([x, aggr_out], dim=1))
        return (self.edge_embed, node_embed)


class Net(pl.LightningModule):
    def __init__(self, dim_node: int = 4, dim_edge: int = 0,
                 dim_embed_edge: int = 64, dim_embed_node: int = 32,
                 num_hidden: int = 3, final_bias: bool = False,
                 pos_weight: float = 80.0,
                 learn_rate: float = 1e-4, weight_decay: float = 5e-4,
                 infer_thresh: float = 0.5):
        super(Net, self).__init__()
        # define the architecture
        self.encode = Interaction(dim_edge, dim_node,
                                  dim_embed_edge, dim_embed_node)
        self.message = pyg.nn.Sequential('x, edge_index, edge_attrs', [
            (Interaction(dim_embed_edge, dim_embed_node,
                         dim_embed_edge, dim_embed_node),
             'x, edge_index, edge_attrs -> edge_attrs, x')
             for i in range(num_hidden)
             ])
        self.classify = torch.nn.Linear(dim_embed_edge, 1, bias=final_bias)
        # optimiser args
        self.lr = learn_rate
        self.decay = weight_decay
        # configure the loss
        self.criterion = torch.nn.BCEWithLogitsLoss(
                pos_weight=torch.tensor(pos_weight, device=self.device),
                reduction='mean')
        # add metrics
        self.train_ACC = torchmetrics.Accuracy(threshold=infer_thresh)
        self.train_F1 = torchmetrics.F1(
                num_classes=1, threshold=infer_thresh)
        self.val_ACC = torchmetrics.Accuracy(threshold=infer_thresh)
        self.val_F1 = torchmetrics.F1(
                num_classes=1, threshold=infer_thresh)
        self.val_PR = torchmetrics.BinnedPrecisionRecallCurve(
                num_classes=1, num_thresholds=5)

    def forward(self, data, sigmoid=True):
        node_attrs, edge_attrs = data.x, data.edge_attr
        edge_attrs, node_attrs = self.encode(node_attrs, data.edge_index,
                                             edge_attrs)
        edge_attrs, node_attrs = self.message(node_attrs, data.edge_index,
                                              edge_attrs)
        pred = self.classify(edge_attrs)
        if sigmoid:
            pred = torch.sigmoid(pred)
        return pred

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(
                self.parameters(),
                lr=self.lr,
                weight_decay=self.decay
            )
        return optimizer

    def _train_av_loss(self, outputs):
        return torch.stack([x['loss'] for x in outputs]).mean()

    def _val_av_loss(self, losses):
        return torch.stack(losses).mean()

    def training_step(self, batch, batch_idx):
        edge_pred = self(batch, sigmoid=False)
        loss = self.criterion(edge_pred, batch.y.view(-1, 1))
        return {'loss': loss,
                'preds': torch.sigmoid(edge_pred),
                'target': batch.y.view(-1, 1).int()}

    def training_step_end(self, outputs):
        self.train_ACC(outputs['preds'], outputs['target'])
        self.train_F1(outputs['preds'], outputs['target'])
        self.log('ptl/train_loss', outputs['loss'], on_step=True)
        return outputs['loss']

    def training_epoch_end(self, outputs):
        self.log('ptl/train_loss', self._train_av_loss(outputs))
        self.log('ptl/train_accuracy', self.train_ACC.compute())
        self.log('ptl/train_f', self.train_F1.compute())

    def validation_step(self, batch, batch_idx):
        edge_pred = self(batch, sigmoid=False)
        loss = self.criterion(edge_pred, batch.y.view(-1, 1))
        return {'loss': loss,
                'preds': torch.sigmoid(edge_pred),
                'target': batch.y.view(-1, 1).int()}

    def validation_step_end(self, outputs):
        self.val_ACC(outputs['preds'], outputs['target'])
        self.val_F1(outputs['preds'], outputs['target'])
        self.val_PR(outputs['preds'], outputs['target'])
        self.log('ptl/val_loss', outputs['loss'], on_step=True)
        return outputs['loss']

    def validation_epoch_end(self, outputs):
        metrics = {
            'ptl/val_loss': self._val_av_loss(outputs),
            'ptl/val_accuracy': self.val_ACC.compute(),
            'ptl/val_f': self.val_F1.compute(),
            }
        self.log_dict(metrics, sync_dist=True)
        prec, recall, thresh = self.val_PR.compute()
        for i, t in enumerate(thresh):
            self.log(f'ptl/val_prec_thresh_{t:.3f}', prec[i])
            self.log(f'ptl/val_recall_thresh_{t:.3f}', recall[i])
        self.trainer.callback_metrics = metrics
        return metrics
```

Here I attach the tuning script using `Ray[Tune]`, as per their docs https://docs.ray.io/en/master/tune/tutorials/tune-pytorch-lightning.html.
```python
import os

import pytorch_lightning as pl
from ray import tune
from ray.tune import CLIReporter
from ray.tune.schedulers import ASHAScheduler
from ray.tune.suggest.hyperopt import HyperOptSearch
from ray.tune.integration.pytorch_lightning import TuneReportCallback
from ray.tune.integration.pytorch_lightning import TuneReportCheckpointCallback

from cluster_gnn.models import gnn
from cluster_gnn.data import loader

# slurm hack
os.environ["SLURM_JOB_NAME"] = "bash"

ROOT_DIR = '/home/jlc1n20/projects/cluster_gnn/'
MODEL_DIR = ROOT_DIR + '/models/tune/'

def train_gnn(config, data_module, num_epochs=10, num_gpus=4, callbacks=None,
              checkpoint_dir=None):
    logger = pl.loggers.TensorBoardLogger(
        save_dir=tune.get_trial_dir(), name="", version=".")
    if checkpoint_dir:
        ckpt = pl.utilities.cloud_io.pl_load(
            os.path.join(checkpoint_dir, 'checkpoint'),
            map_location=lambda storage, loc: storage)
        model = gnn.Net._load_model_state(
            checkpoint=ckpt,
            num_hidden=6, dim_embed_edge=64, dim_embed_node=32,
            learn_rate=config['learn_rate'],
            pos_weight=config['pos_weight'])
    else:
        model = gnn.Net(num_hidden=6, dim_embed_edge=64, dim_embed_node=32,
                        learn_rate=config['learn_rate'],
                        pos_weight=config['pos_weight'])
    trainer = pl.Trainer(gpus=num_gpus, num_nodes=1, max_epochs=num_epochs,
                         progress_bar_refresh_rate=0,
                         limit_train_batches=0.1,
                         logger=logger,
                         callbacks=callbacks)
    trainer.fit(model, data_module)
    print('callback metrics are:\n {}'.format(trainer.callback_metrics))

def tune_gnn(data_module, num_samples=10, num_epochs=10, gpus_per_trial=2,
             init_params=None, checkpoint_dir=None):
    config = {
        'learn_rate': tune.loguniform(1e-6, 1e-1),
        'pos_weight': tune.uniform(1.0, 100.0),
        }
    metrics = ['ptl/val_loss', 'ptl/val_accuracy', 'ptl/val_f']
    callbacks = [
        TuneReportCheckpointCallback(
            metrics,
            filename='checkpoint',
            on='validation_end')
        ]
    scheduler = ASHAScheduler(
        time_attr='epoch',
        max_t=num_epochs,
        )
    search_alg = HyperOptSearch(points_to_evaluate=init_params)
    reporter = CLIReporter(
        parameter_columns=[
            'learn_rate',
            'pos_weight',
            ],
        )
    trainable = tune.with_parameters(
        train_gnn,
        data_module=data_module,
        num_epochs=num_epochs,
        num_gpus=gpus_per_trial,
        callbacks=callbacks,
        checkpoint_dir=checkpoint_dir,
        )
    analysis = tune.run(
        trainable,
        resources_per_trial={
            'cpu': 1,
            'gpu': gpus_per_trial,
            },
        metric='ptl/val_f',
        mode='max',
        config=config,
        num_samples=num_samples,
        search_alg=search_alg,
        scheduler=scheduler,
        progress_reporter=reporter,
        local_dir=MODEL_DIR,
        verbose=3,
        name='tune_gnn')
    print('Best hp found: ', analysis.best_config)


if __name__ == '__main__':
    num_gpus = 2
    cur_best_params = [{
        'learn_rate': 3.75e-5,
        'pos_weight': 21.5,
        }]
    graph_data = loader.GraphDataModule(
        '/home/jlc1n20/projects/cluster_gnn/data/', num_workers=num_gpus)

    tune_gnn(data_module=graph_data, num_samples=1, num_epochs=1,
             gpus_per_trial=num_gpus, init_params=cur_best_params)

```

### PR DESCRIPTION
**Pull request by edgarriba**
## What does this PR do?

Fixes #7671 
Fixes access to callback_metrics in ddp_spawn

**TODO:**

- [ ] tests
- [ ] docs

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 1]
```diff
@@ -270,6 +271,10 @@ def determine_ddp_device_ids(self):
             return None
         return [self.root_device.index]
 
+    from typing import Dict
+    def _dict_to_float(self, dict_tensor: Dict[str, torch.Tensor]) -> Dict[str, float]:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:276]: You can use apply_to_collection.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 2]
```diff
@@ -290,6 +295,7 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
             self.mp_queue.put(best_model_path)
             self.mp_queue.put(last_path)
             self.mp_queue.put(results)
+            self.mp_queue.put(self._dict_to_float(self.lightning_module.trainer.logger_connector.callback_metrics))
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:299]: I believe this should be something the user can select by himself as a property or hook.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py [hunk 3]
```diff
@@ -277,6 +278,14 @@ def reset(self, metrics: Optional[bool] = None) -> None:
         self._batch_idx = None
         self._split_idx = None
         self._current_fx = None
+    
+    @property
+    def spawn_callback_metrics(self) -> Dict[MetricSource, Dict[str, _METRIC]]:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:283]: I think this should be more general than just callback_metrics.

``spawn_extra_parameters``.

**Review by carmocca — COMMENTED** [review]
Should we put the callback metrics directly in the queue instead?
We don't want users to have to use a different attribute depending on the accelerator.

How would this impact performance?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py (hunk 3 above)
Comment (edgarriba) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector_new.py:283]: done

**Comment by edgarriba** [comment]
@carmocca My initial proposal using the `optuna` framework as entry point example is the following:
```python
return trainer.spawn_callback_metrics["val_acc"]
```
however, as @tchaton proposes to make it more generic we could follow the approach below
```python
return trainer.spawn_extra_parameters["callback_metrics"]["val_acc"]
```
open for an api discussion

** this is the gist to entry point script: https://gist.github.com/edgarriba/af6247edb32586b19e740f17735ff055

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 4]
```diff
@@ -214,6 +215,8 @@ def post_dispatch(self):
         best_path = self.mp_queue.get()
         last_path = self.mp_queue.get()
         self._results = self.mp_queue.get()
+        # TODO: verify to find corner cases or concurrency issues
+        self.lightning_module.trainer.logger_connector.spawn_extra_parameters.update(self.mp_queue.get())
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:219]:
```suggestion
        self._results = self.mp_queue.get()
        extra_parameters = self.mp_queue.get()
        # TODO: verify to find corner cases or concurrency issues
        self.lightning_module.trainer.logger_connector.spawn_extra_parameters.update(extra_parameters)
```
for visual clarity, all get() from queue listed line by line :)

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 5]
```diff
@@ -290,6 +293,11 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
             self.mp_queue.put(best_model_path)
             self.mp_queue.put(last_path)
             self.mp_queue.put(results)
+            self.mp_queue.put({
+                "callback_metrics": apply_to_collection(
+                    self.lightning_module.trainer.logger_connector.callback_metrics, torch.Tensor, lambda x: x.item()
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:298]: do you know why we need to call x.item()?

**Review by awaelchli — COMMENTED** [review]
some initial comments

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 5 above)
Comment (edgarriba) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:299]: there was an issue when pickling cuda tensors. We discussed with @tchaton to cast tensors to primitive float types

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 6]
```diff
@@ -214,6 +215,9 @@ def post_dispatch(self):
         best_path = self.mp_queue.get()
         last_path = self.mp_queue.get()
         self._results = self.mp_queue.get()
+        extra_parameters = self.mp_queue.get()
+        # TODO: verify to find corner cases or concurrency issues
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:219]: what kind of issues are you thinking here? will you resolve the todo in this PR

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:299]: and doing .cpu() instead of .item(), have you thought about it?

File: tests/plugins/test_ddp_spawn_plugin.py [hunk 7]
```diff
@@ -40,3 +52,41 @@ def test_ddp_cpu():
...
+    model = BoringCallbackDDPSpawnModel(val_name, val)
+    dm = BoringDataModule()
+
+    trainer.fit(model, datamodule=dm)
+    assert "callback_metrics" in trainer.spawn_extra_parameters
+    assert trainer.spawn_extra_parameters["callback_metrics"][val_name] == val
+
+
+@RunIf(min_gpus=2)
+def test_ddp_multigpu():
```
Comment (awaelchli) [inline tests/plugins/test_ddp_spawn_plugin.py:77]: this test is identical to the above with the exception of the gpu setting. 
Can we just do the 2 gpu case? I don't see the value for single gpu test

File: tests/plugins/test_ddp_spawn_plugin.py [hunk 8]
```diff
@@ -40,3 +52,41 @@ def test_ddp_cpu():
...
+    """Tests if device is set correctely when training for DDPSpawnPlugin."""
+    trainer = Trainer(num_processes=2, fast_dev_run=True, gpus=2, accelerator='ddp_spawn')
```
Comment (awaelchli) [inline tests/plugins/test_ddp_spawn_plugin.py:79]:
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, gpus=2, accelerator="ddp_spawn")
```
num_processes would only be needed for ddp_cpu

File: tests/plugins/test_ddp_spawn_plugin.py (hunk 7 above)
Comment (awaelchli) [inline tests/plugins/test_ddp_spawn_plugin.py:77]:
```suggestion
def test_ddp_spawn_extra_parameters(tmpdir):
```

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 9]
```diff
@@ -290,6 +294,11 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
             self.mp_queue.put(best_model_path)
             self.mp_queue.put(last_path)
             self.mp_queue.put(results)
+            self.mp_queue.put({
+                "callback_metrics": apply_to_collection(
+                    self.lightning_module.trainer.logger_connector.callback_metrics, torch.Tensor, lambda x: x.item()
+                ) or {}
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:300]: Why do we need `or {}`? Should return an empty dict by default if I'm not mistaken

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 10]
```diff
@@ -277,6 +278,14 @@ def reset(self, metrics: Optional[bool] = None) -> None:
         self._split_idx = None
         self._current_fx = None
 
+    @property
+    def spawn_extra_parameters(self) -> Dict[str, Dict[str, _METRIC]]:
+        return self._spawn_extra_parameters
+
+    @spawn_extra_parameters.setter
+    def spawn_extra_parameters(self, val: Dict[str, Dict[str, _METRIC]]) -> None:
+        self._spawn_extra_parameters = val
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:287]: Since these are not customizing anything, you can have the property directly in the `__init__`

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 5 above)
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:299]: @edgarriba I also suggest using a cpu-tensor instead of primitives if possible. This means that we can still use PyTorch's multiprocessing package (depending on shared memory instead of pickling) and we avoid the tensor creation on the main proc (IMO the type has to be restored otherwise)

**Comment by tchaton** [comment]
Hey @carmocca @awaelchli,

Should we move this property onto the LightningModule. Users could override it to return anything they want.

Best,
T.C

**Comment by kaushikb11** [comment]
@edgarriba This would affect TPU Spawn as well, hence the tests are failing. (The hang due to extra parameters being expected by the queue)

**Comment by carmocca** [comment]
> Should we move this property onto the LightningModule. Users could override it to return anything they want.

`spawn_extra_parameters`? Why would they want to?

**Comment by tchaton** [comment]
> > Should we move this property onto the LightningModule. Users could override it to return anything they want.
> 
> `spawn_extra_parameters`? Why would they want to?

Hey @carmocca.

As a user, I would like to be able to return anything directly from the queue such as Callback, predictions, or anything else.

I had something like that in mind.

```py
class LightningModule

    self._spawn_extra_parameters = {}

    @property
    def spawn_extra_parameters(self):
        return self._spawn_extra_parameters

    @spawn_extra_parameters.setter
    def spawn_extra_parameters(self, spawn_extra_parameters):
        self.spawn_extra_parameters = spawn_extra_parameters


class Model(LightningModule):

    def add_spawn_extra_parameters(self):
        return {"callbacks_metrics": self.trainer.callback_metrics, "state": self.trainer....}


model = Model()
trainer = Trainer(gpus=2, accelerator="ddp_spawn")
trainer.fit(model)
model.spawn_extra_parameters["callbacks_metrics"] # my callback metrics from process 0
model.spawn_extra_parameters["state"] # my state from process 0

```

**Comment by carmocca** [comment]
That could be useful but I still think that we shouldn't introduce a new property for it

> We don't want users to have to use a different attribute depending on the accelerator.

What if we use `callback_metrics` directly? A big part of Lightning is not having to change your code when changing training types.

**Comment by edgarriba** [comment]
@kaushikb11 TPU issue seems fixed.

**Comment by tchaton** [comment]
> That could be useful but I still think that we shouldn't introduce a new property for it
> 
> > We don't want users to have to use a different attribute depending on the accelerator.
> 
> What if we use `callback_metrics` directly? A big part of Lightning is not having to change your code when changing training types.

It is not accelerator dependent, by process creation dependent. 
I think Lightning should enable users to provide a hook for the ≈.

 ```py
def add_to_queue(self, queue):
     queue.put(self.trainer.callback_metrics)

def get_from_queue(self, queue):
    self.trainer.callback_metrics = queue.get()
```

File: pytorch_lightning/core/lightning.py [hunk 11]
```diff
@@ -1948,3 +1949,15 @@ def model_size(self) -> float:
         size_mb = os.path.getsize(tmp_name) / 1e6
         os.remove(tmp_name)
         return size_mb
+
+    def add_to_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
```
Comment (edgarriba) [inline pytorch_lightning/core/lightning.py:1953]: @tchaton @carmocca should we make this more generic like ?
```python
def add_hook_to_queue(self, queue: torch.multiprocessing.SimpleQueue, hook_name: str) -> None:
   data = getattr(self.trainer, hook_name)
   queue.put(data)
```
then 
```python
def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue, hook_name: str) -> None:
   data = queue.get()
   setattr(self.trainer, hook_name, data)
```

Comment (carmocca) [reply pytorch_lightning/core/lightning.py:1953]: I don't think so. Do you have any examples where that would be particularly useful?

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1953]: Hey @edgarriba,

This won't work. User might want to do the random data.

```
# Lightning Module

def add_data_to_spawn_queue(self) -> Any:
    queue.put(self.trainer.callback_metrics)

def get_data_to_spawn_queue(self, data):
    # do something with the data.

# DDP Spawn

self.mq_queue.add(self.lightning_module. add_data_to_spawn_queue())

self.lightning_module. get_data_to_spawn_queue(self.mq_queue.get())
```

Comment (edgarriba) [reply pytorch_lightning/core/lightning.py:1953]: not right now, but in case is needed can easily implemented following this pattern.

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1953]: @awaelchli @carmocca ?

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 12]
```diff
@@ -290,6 +297,14 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
             self.mp_queue.put(best_model_path)
             self.mp_queue.put(last_path)
             self.mp_queue.put(results)
+            self.lightning_module.add_to_queue(self.mp_queue)
+            '''self.mp_queue.put({
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:301]: Same here. This description should be within the function description and only a small comment should be placed above.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 13]
```diff
@@ -202,6 +202,14 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
                 self.mp_queue.put(best_model_path)
                 self.mp_queue.put(last_path)
                 self.mp_queue.put(results)
+                self.lightning_module.add_to_queue(self.mp_queue)
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:205]: Same here

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 14]
```diff
@@ -277,6 +278,14 @@ def reset(self, metrics: Optional[bool] = None) -> None:
         self._split_idx = None
         self._current_fx = None
 
+    '''@property
+    def spawn_extra_parameters(self) -> Dict[str, Dict[str, _METRIC]]:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:282]: same here

File: pytorch_lightning/core/lightning.py (hunk 11 above)
Comment (justusschock) [reply pytorch_lightning/core/lightning.py:1953]: I think we should only care for `callback_metrics` right now. Should be easy to extend once anything else is requested anyways. Also I think in case of spawn we should definitely forward the `callback_metrics` property to this thing to prevent code changes for the user (as @carmocca suggested). 

@tchaton you said that this isn't accelerator specific and you're right, but if possible we definitely should also avoid code changes for different training types.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 15]
```diff
@@ -214,6 +216,11 @@ def post_dispatch(self):
         best_path = self.mp_queue.get()
         last_path = self.mp_queue.get()
         self._results = self.mp_queue.get()
+        self.lightning_module.get_from_queue(self.mp_queue)
+        '''extra_parameters = self.mp_queue.get()
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:220]: this is the old code, pending to be removed after the final api decision. Same for the others

File: pytorch_lightning/core/lightning.py [hunk 16]
```diff
@@ -1948,3 +1949,30 @@ def model_size(self) -> float:
...
+        """Appends the `trainer.callback_metrics` dictionary to the given queue.
+
+        To avoid issues with memory sharing, we cast the data to numpy arrays.
+
+        Args:
+            queue: the instance of the queue to append the data.
+
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1960]:
```suggestion
```

File: pytorch_lightning/core/lightning.py [hunk 17]
```diff
@@ -1948,3 +1949,30 @@ def model_size(self) -> float:
...
+        queue.put(callback_metrics)
+
+    def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
+        """Retrieve the `trainer.callback_metrics` dictionary from the given queue.
+
+        To preserve consistency with torch.Tensor, we cast back the data to tensors.
+
+        Args:
+            queue: the instance of the queue from where to get the data.
+
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1974]:
```suggestion
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 18]
```diff
@@ -290,6 +290,10 @@ def callback_metrics(self) -> Dict[str, _METRIC]:
             self._callback_metrics.update(metrics)
         return self._callback_metrics
 
+    @callback_metrics.setter
+    def callback_metrics(self, val: dict) -> None:
+        self._callback_metrics = val
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:296]: Shouldn't need this, you can update the dictionary instead

File: pytorch_lightning/trainer/trainer.py [hunk 19]
```diff
@@ -537,6 +537,7 @@ def validate(
         self.state.fn = TrainerFn.VALIDATING
         self.state.status = TrainerStatus.RUNNING
         self.validating = True
+        self.validated_ckpt_path: Optional[str] = None
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:540]: These are already defined in `EvaluationLoop.on_trainer_init`.

Also would be nice for you to split the typing fixes into a separate PR

File: tests/plugins/test_ddp_spawn_plugin.py [hunk 20]
```diff
@@ -26,6 +26,18 @@ def on_train_start(self) -> None:
         assert self.device == torch.device("cpu")
 
 
+class BoringCallbackDDPSpawnModel(BoringModel):
```
Comment (carmocca) [inline tests/plugins/test_ddp_spawn_plugin.py:29]: could override the `{get,add}_to_queue` hook and add something to check

File: pytorch_lightning/core/lightning.py [hunk 21]
```diff
@@ -1948,3 +1949,30 @@ def model_size(self) -> float:
...
+            queue: the instance of the queue to append the data.
+
+        """
+        callback_metrics: dict = apply_to_collection(
+            self.trainer.callback_metrics, torch.Tensor, lambda x: x.cpu().numpy()
+        )  # send as numpy to avoid issues with memory sharing
+        queue.put(callback_metrics)
+
+    def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
+        """Retrieve the `trainer.callback_metrics` dictionary from the given queue.
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1968]: Use `:attr:` for callback metrics

File: pytorch_lightning/core/lightning.py [hunk 22]
```diff
@@ -1948,3 +1949,30 @@ def model_size(self) -> float:
...
+
+        To preserve consistency with torch.Tensor, we cast back the data to tensors.
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1970]:
```suggestion
        To preserve consistency, we cast back the data to ``torch.Tensor``.
```

File: pytorch_lightning/core/lightning.py [hunk 23]
```diff
@@ -1948,3 +1949,30 @@ def model_size(self) -> float:
         size_mb = os.path.getsize(tmp_name) / 1e6
         os.remove(tmp_name)
         return size_mb
+
+    def add_to_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
+        """Appends the `trainer.callback_metrics` dictionary to the given queue.
+
+        To avoid issues with memory sharing, we cast the data to numpy arrays.
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1956]:
```suggestion
        To avoid issues with memory sharing, we cast the data to numpy.
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 18 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:296]: +1 same for Trainer.callback_metrics who forwards here.

**Review by tchaton — APPROVED** [review]
Overall looks good.

**Comment by awaelchli** [comment]
Will you also add the new public methods to the docs? i.e., under the LightningModule api reference?

**Comment by edgarriba** [comment]
@carmocca @awaelchli your comments I believe that were adressed

**Review by awaelchli — APPROVED** [review]
great
I think just the changelog entries missing now. 
Adding 1.4 milestone

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 24]
```diff
@@ -1948,3 +1949,30 @@ def model_size(self) -> float:
...
+
+    def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
+        """Retrieve the :attr:`trainer.callback_metrics` dictionary from the given queue.
+
+        To preserve consistency, we cast back the data to ``torch.Tensor``.
+
+        Args:
+            queue: the instance of the queue from where to get the data.
+        """
+        # NOTE: this must be called in the right order to get the `callback_metrics`
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1974]: Is this what you meant? Could put it in the docstring

```suggestion
        # NOTE: `add_to_queue` needs to be called before
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 25]
```diff
@@ -1948,3 +1949,30 @@ def model_size(self) -> float:
...
+        To preserve consistency, we cast back the data to ``torch.Tensor``.
+
+        Args:
+            queue: the instance of the queue from where to get the data.
+        """
+        # NOTE: `add_to_queue` needs to be called before
+        callback_metrics: dict = queue.get()
+        self.trainer.callback_metrics.update(
+            apply_to_collection(callback_metrics, np.ndarray, lambda x: torch.tensor(x))
+        )
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1978]: is this the only alternative to populate these metrics? why is this on a user the user interace of the lightning module? what happens if someone overrides this? is it meant to be overridden? 

it feels like the lightning module is used as a go-between between different parts of the trainer, in particular because the training type plugin technically has no reference to the trainer. 

structrually, we are repeatedly reaching through the lightning module to access the trainer in a very roundabout way. another example: https://github.com/PyTorchLightning/pytorch-lightning/blob/55a90af7fc0805855684e93dbad669f5bbe76eee/pytorch_lightning/plugins/training_type/sharded.py#L42-L57

it feels backwards and it also makes efforts like https://github.com/PyTorchLightning/pytorch-lightning/issues/7315 harder to work through when we keep relying on the trainer like this

**Comment by awaelchli** [comment]
Do `add_to_queue` and `get_from_queue` need to be public? @ananthsub suggests them to be protected.

**Comment by carmocca** [comment]
> Do `add_to_queue` and `get_from_queue` need to be public?

The point was to let users add and remove from these. See https://github.com/PyTorchLightning/pytorch-lightning/pull/7916#issuecomment-860583628 and the rest of the discussions in this PR
